File: backend/app/schemas/recurring.py
"""
定期交易 Schema
"""
import uuid
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

from app.schemas.transaction import TransactionCreateRequest


RecurringInterval = Literal["daily", "weekly", "monthly", "yearly"]


class RecurringTemplateData(TransactionCreateRequest):
    transaction_date: None = None


class RecurringCreateRequest(BaseModel):
    interval: RecurringInterval
    next_run_date: date
    template_data: RecurringTemplateData


class RecurringUpdateRequest(BaseModel):
    interval: Optional[RecurringInterval] = None
    next_run_date: Optional[date] = None
    is_active: Optional[bool] = None
    template_data: Optional[RecurringTemplateData] = None


class RecurringResponse(BaseModel):
    id: uuid.UUID
    ledger_id: uuid.UUID
    created_by: Optional[uuid.UUID]
    interval: str
    next_run_date: date
    is_active: bool
    template_data: dict
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/recurring.ts
import { apiClient } from '@/api'
import type { Necessity, TransactionCreatePayload } from '@/api/transactions'

export type RecurringInterval = 'daily' | 'weekly' | 'monthly' | 'yearly'

export interface RecurringTemplateData extends Omit<TransactionCreatePayload, 'transaction_date'> {
  amount: number
  currency_code?: string
  necessity?: Necessity
}

export interface RecurringTemplate {
  id: string
  ledger_id: string
  created_by: string | null
  interval: RecurringInterval
  next_run_date: string
  is_active: boolean
  template_data: RecurringTemplateData
  created_at: string
  updated_at: string
}

export interface RecurringCreatePayload {
  interval: RecurringInterval
  next_run_date: string
  template_data: RecurringTemplateData
}

export type RecurringUpdatePayload = Partial<RecurringCreatePayload> & {
  is_active?: boolean
}

export async function listRecurringTemplates(ledgerId: string): Promise<RecurringTemplate[]> {
  const response = await apiClient.get<RecurringTemplate[]>(`/ledgers/${ledgerId}/recurring`)
  return response.data
}

export async function createRecurringTemplate(
  ledgerId: string,
  payload: RecurringCreatePayload,
): Promise<RecurringTemplate> {
  const response = await apiClient.post<RecurringTemplate>(`/ledgers/${ledgerId}/recurring`, payload)
  return response.data
}

export async function updateRecurringTemplate(
  ledgerId: string,
  recurringId: string,
  payload: RecurringUpdatePayload,
): Promise<RecurringTemplate> {
  const response = await apiClient.patch<RecurringTemplate>(`/ledgers/${ledgerId}/recurring/${recurringId}`, payload)
  return response.data
}

export async function deleteRecurringTemplate(ledgerId: string, recurringId: string): Promise<void> {
  await apiClient.delete(`/ledgers/${ledgerId}/recurring/${recurringId}`)
}
